File: web/src/screen.ts
export abstract class Screen {
  abstract onDraw(ctx: CanvasRenderingContext2D): void;
  abstract onLogic(): void;

  onEntered(): void {}
  onExited(): void {}
  onEnter(_ctx: CanvasRenderingContext2D): boolean { return true; }
  onExit(_ctx: CanvasRenderingContext2D): boolean { return true; }

  exit(): void {
    ScreenManager.exit(this);
  }

  isTop(): boolean {
    return ScreenManager.getTop() === this;
  }
}

export const ScreenManager = {
  _stack: [] as Screen[],
  _screenToEnter: null as Screen | null,
  _screenToExit: null as Screen | null,

  add(screen: Screen): void {
    this._stack.push(screen);
    this._screenToEnter = screen;
  },

  exit(screen: Screen): void {
    if (this.getTop() !== screen) return;
    if (this._screenToExit !== null) return;
    this._screenToExit = screen;
  },

  getTop(): Screen | null {
    return this._stack.length > 0 ? this._stack[this._stack.length - 1]! : null;
  },

  isEmpty(): boolean { return this._stack.length === 0; },

  onLogic(): void {
    if (this._screenToExit === null && this._screenToEnter === null) {
      this.getTop()?.onLogic();
    }
  },

  draw(ctx: CanvasRenderingContext2D): void {
    if (this._screenToExit === null && this._screenToEnter === null) {
      this.getTop()?.onDraw(ctx);
    }

    if (this._screenToExit !== null) {
      const exitDone = this._screenToExit.onExit(ctx);
      if (!exitDone) return;
      const idx = this._stack.indexOf(this._screenToExit);
      if (idx >= 0) this._stack.splice(idx, 1);
      this._screenToExit.onExited();
      this._screenToExit = null;
      if (this.getTop() !== null) this._screenToEnter = this.getTop();
    }

    if (this._screenToEnter !== null) {
      const enterDone = this._screenToEnter.onEnter(ctx);
      if (!enterDone) return;
      this._screenToEnter.onEntered();
      this._screenToEnter = null;
    }
  },
};

File: wii/src/screen.cpp
#include "screen.h"
#include <algorithm>

void Screen::exit() {
    ScreenManager::exit(this);
}

bool Screen::isTop() const {
    return ScreenManager::getTop() == this;
}

void ScreenManager::add(Screen* screen) {
    stack_.push_back(screen);
    screenToEnter_ = screen;
}

void ScreenManager::exit(Screen* screen) {
    if (getTop() != screen) return;
    if (screenToExit_ != nullptr) return;
    screenToExit_ = screen;
}

Screen* ScreenManager::getTop() {
    return stack_.empty() ? nullptr : stack_.back();
}

bool ScreenManager::isEmpty() {
    return stack_.empty();
}

void ScreenManager::onLogic() {
    if (screenToExit_ == nullptr && screenToEnter_ == nullptr) {
        auto* top = getTop();
        if (top) top->onLogic();
    }
}

void ScreenManager::draw(SDL_Renderer* renderer) {
    if (screenToExit_ == nullptr && screenToEnter_ == nullptr) {
        auto* top = getTop();
        if (top) top->onDraw(renderer);
    }

    if (screenToExit_ != nullptr) {
        bool exitDone = screenToExit_->onExit(renderer);
        if (!exitDone) return;
        auto it = std::find(stack_.begin(), stack_.end(), screenToExit_);
        if (it != stack_.end()) {
            stack_.erase(it);
        }
        screenToExit_->onExited();
        delete screenToExit_;
        screenToExit_ = nullptr;
        if (getTop() != nullptr) screenToEnter_ = getTop();
    }

    if (screenToEnter_ != nullptr) {
        bool enterDone = screenToEnter_->onEnter(renderer);
        if (!enterDone) return;
        screenToEnter_->onEntered();
        screenToEnter_ = nullptr;
    }
}

void ScreenManager::clear() {
    for (auto* s : stack_) delete s;
    stack_.clear();
    screenToEnter_ = nullptr;
    screenToExit_ = nullptr;
}
